src/sentry/eventstream/errors/backend.py

```diff
@@ -0,0 +1,323 @@
+        commit_log_topic: str,
+        synchronize_commit_group: str,
+        commit_batch_size: int = 100,
+        commit_batch_timeout_ms: int = 5000,
+        initial_offset_reset: Union[Literal["latest"], Literal["earliest"]] = "latest",
+    ) -> bool:
+        self._backend.run_forwarder(
+            consumer_group=consumer_group,
+            commit_log_topic=commit_log_topic,
+            synchronize_commit_group=synchronize_commit_group,
+            commit_batch_size=commit_batch_size,
+            commit_batch_timeout_ms=commit_batch_timeout_ms,
+            initial_offset_reset=initial_offset_reset,
+        )
+
+    def requires_post_process_forwarder(self):
+        return self._backend.requires_post_process_forwarder
+
+    def _get_extra_data(self, data: ErrorsInsertData, project, event_data) -> Mapping:
+        retention_days = quotas.get_event_retention(organization=project.organization)
+        event = data.event
+        return (
+            {
+                "group_id": event.group_id,
+                "event_id": event.event_id,
+                "organization_id": project.organization_id,
+                "project_id": event.project_id,
+                # TODO(mitsuhiko): We do not want to send this incorrect
+                # message but this is what snuba needs at the moment.
+                "message": event.search_message,
+                "platform": event.platform,
+                "datetime": event.datetime,
+                "data": event_data,
+                "primary_hash": data.primary_hash,
+                "retention_days": retention_days,
+            },
+            {
+                "is_new": data.is_new,
+                "is_regression": data.is_regression,
+                "is_new_group_environment": data.is_new_group_environment,
+                "skip_consume": data.skip_consume,
+            },
+        )
+
+    def insert(self, data: ErrorsInsertData) -> None:
+        project = data.event.project
+        set_current_event_project(project.id)
+        event_data = data.event.get_raw_data(for_stream=True)
+
+        unexpected_tags = get_unexpected_tags(event_data)
+        if unexpected_tags:
+            logger.error("%r received unexpected tags: %r", self, unexpected_tags)
+
+        headers = strip_none_values(
+            {
+                "Received-Timestamp": str(data.received_timestamp),
+                "event_id": str(data.event.event_id),
+                "project_id": str(data.event.project_id),
+                "group_id": str(data.event.group_id),
+                "primary_hash": str(data.primary_hash),
+                "is_new": encode_bool(data.is_new),
+                "is_new_group_environment": encode_bool(data.is_new_group_environment),
+                "is_regression": encode_bool(data.is_regression),
+                "skip_consume": encode_bool(data.skip_consume),
+                "transaction_forwarder": encode_bool(True),
+            }
+        )
```

Comment:
Related to the comment above about the generic. 
It seems that all the event type specific information you need here is in order to generate headers. 
A simple idea could be to have an abstract class on top of `ErrorsInsertData` and `TransactionsInsertData` and have a `get_headers` method there that generates headers. Now there would be no need to expose the details anymore to the EventStreamAPI and no need for the generic. 

Though that has downsides:
You would be creating two parallel hierarchies. One for EventStreamAPI and one for the relevant types. This is not necessarily bad but it smells over complication. 
This over-complication comes from the fact that, in order to use `EventStreamAPI` you would need to know the specific instance you need and the specific object you need to pass. Which means that, in order to call the API you need to know a lot of the implementation details of each implementation. 
This makes the abstract class pointless as it does not abstract anything away.  
Another signal of this is the fact that all implementation of the API are used in the same place (save_event) and not anymore.

So i would reconsider the idea of not having an abstract `EventStreamAPI` or to have an abstract class only on top of `run_post_process_forwarder` which seems more abstract.